graphql/ratelimit.go

```diff
@@ -0,0 +1,172 @@
+package graphql
+
+import (
+	"sync"
+	"time"
+
+	"github.com/satori/go.uuid"
+)
+
+type endRequestState int
+
+const (
+	maxSimultaneousRequestsDefault int             = 15
+	minSimultaneousRequestsDefault int             = 2
+	waitTimeDefault                time.Duration   = 3 * time.Second
+	endRequestStateOK              endRequestState = 0
+	endRequestStateError           endRequestState = 1
+	endRequestStateTimedOut        endRequestState = 2
+)
+
+// RatelimitObject represents structure of ratelimit object
+type RatelimitObject struct {
+	// requestsBucket a list of all simultaneous requests
+	requestsBucket []activeRequest
+	// maxRequests the max possible amount of simultaneous requests that will
+	// have service
+	maxRequests int
+	// minRequests the min possible amount of simultaneous requests
+	minRequests int
+	// currentMaxRequestsLevel an amount of simultaneous requests that could be
+	// processed in current situation. This count is depends on timeouted
+	// requests to databases
+	currentMaxRequestsLevel int
+	// predictedDuration a duration that could take current request. Calculation
+	// depends on real duration of previous requests
+	predictedDuration time.Duration
+	// if simultaneous requests of service reach level of
+	// currentMaxRequestsLevel service will wait `waitTime` before repeating
+	// attempt to serve request
+	waitTime time.Duration
+	mux      sync.Mutex
+}
+
+// activeRequest provides structure for request that processing by service
+// stores id, count of simultaneous requests on moment of starting request,
+// date of request start and predicted request duration
+type activeRequest struct {
+	id          uuid.UUID
+	bucketLevel int
+	startedAt   time.Time
+	predictedAt time.Time
+}
+
+// RatelimitHandlerDefault creates ratelimit object with empty values
+func RatelimitHandlerDefault() *RatelimitObject {
+	rObj := &RatelimitObject{}
+	return rObj.initiate()
+}
+
+// RatelimitHandler creates ratelimit struct object
+func RatelimitHandler(maxRequests int, minRequests int, waitTime time.Duration) *RatelimitObject {
+	rObj := RatelimitObject{
+		maxRequests:             maxRequests,
+		minRequests:             minRequests,
+		currentMaxRequestsLevel: maxRequests,
+		waitTime:                waitTime,
+		predictedDuration:       waitTime,
+	}
+	return rObj.initiate()
+}
+
+// initiate helps initialize ratelimitObject with default values
+func (rObj *RatelimitObject) initiate() *RatelimitObject {
+	rObj.mux.Lock()
```

Comment:
Looks like this function is only used in the constructor, which guarantees it is not accessed concurrently. No need for mutex here.